File: sigilgateapp/adapters/inmemory_etcd.py
class InMemoryEtcd:
    def __init__(self) -> None:
        self._store: dict[str, str] = {}

    def get(self, key: str) -> str | None:
        return self._store.get(key)

    def put(self, key: str, value: str) -> None:
        self._store[key] = value

    def delete(self, key: str) -> None:
        self._store.pop(key, None)

    def prefix_scan(self, prefix: str) -> dict[str, str]:
        return {k: v for k, v in self._store.items() if k.startswith(prefix)}

    def txn(self, ops: list[tuple[str, str | None]]) -> None:
        for key, value in ops:
            if value is None:
                self._store.pop(key, None)
            else:
                self._store[key] = value

    def compare_and_swap(self, key: str, expected: str | None, value: str) -> bool:
        if self._store.get(key) != expected:
            return False
        self._store[key] = value
        return True

File: internal/adapter/inmemory_etcd.go
package adapter

import (
	"context"
	"github.com/SigilGate/SigilGateAPI/internal/port"
	"strings"
    "sync"
)

type InMemoryETCD struct {
	mu sync.Mutex
	store map[string]string
}

var _ port.EtcdPort = (*InMemoryETCD)(nil)

func NewInMemoryETCD() *InMemoryETCD {
	return &InMemoryETCD{store: make(map[string]string)}
}

func (e *InMemoryETCD) Get(_ context.Context, key string) (*string, error) {
	e.mu.Lock()
	defer e.mu.Unlock()
	
	v, ok :=  e.store[key]
	if !ok {
		return nil, nil
	}
	return &v, nil
}

func (e *InMemoryETCD) Put(_ context.Context, key, value string) error {
	e.mu.Lock()
	defer e.mu.Unlock()

	e.store[key] = value
	return nil
}

func (e *InMemoryETCD) Delete(_ context.Context, key string) error {
	e.mu.Lock()
	defer e.mu.Unlock()

	delete(e.store, key)
	return nil
}

func (e *InMemoryETCD) PrefixScan(_ context.Context, prefix string) (map[string]string, error) {
	e.mu.Lock()
	defer e.mu.Unlock()

    m := make(map[string]string)
	for k, v := range e.store {
		if strings.HasPrefix(k, prefix) {
			m[k] = v
		}
	}
	return m, nil
}

func (e *InMemoryETCD) Txn(_ context.Context, ops []port.TxnOp) error {
	e.mu.Lock()
	defer e.mu.Unlock()
	for _, op := range ops {
		if op.Value==nil {
			delete(e.store, op.Key)
		} else {
			e.store[op.Key] = *op.Value
		}
	}
	return nil
}

func (e *InMemoryETCD) CompareAndSwap(_ context.Context, key string, expected *string, value string) (bool, error) {
	e.mu.Lock()
	defer e.mu.Unlock()

    current, exists := e.store[key]
	if expected==nil {
		if exists {
			return false, nil
		}
	} else {
		if !exists || current != *expected {
			return false, nil
		}
	}
    e.store[key] = value
	return true, nil
}
